synapse/visibility.py

```diff
@@ -84,7 +84,14 @@ async def filter_events_for_client(
     """
     # Filter out events that have been soft failed so that we don't relay them
     # to clients.
+    events_before_filtering = events
     events = [e for e in events if not e.internal_metadata.is_soft_failed()]
+    if len(events_before_filtering) != len(events):
+        logger.debug(
+            "filter_events_for_client: Filtered out soft-failed events: Before=%s, After=%s",
+            [event.event_id for event in events_before_filtering],
+            [event.event_id for event in events],
+        )
```

Comment:
We should definitely add a `if logger.isEnabledFor(logging.DEBUG):` guard here so we don't pay the cost of creating the lists when DEBUG logging is disabled